File: src/main/java/ac/kr/deu/connect/luck/auth/service/AuthService.java
package ac.kr.deu.connect.luck.auth.service;

import ac.kr.deu.connect.luck.auth.JwtTokenProvider;
import ac.kr.deu.connect.luck.auth.controller.request.LoginRequest;
import ac.kr.deu.connect.luck.auth.controller.request.SignUpRequest;
import ac.kr.deu.connect.luck.auth.controller.response.EmailResponse;
import ac.kr.deu.connect.luck.auth.controller.response.IdCheckResponse;
import ac.kr.deu.connect.luck.auth.controller.response.TokenResponse;
import ac.kr.deu.connect.luck.exception.CustomErrorCode;
import ac.kr.deu.connect.luck.exception.CustomException;
import ac.kr.deu.connect.luck.user.entity.User;
import ac.kr.deu.connect.luck.user.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class AuthService {
    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider jwtTokenProvider;

    /**
     * 회원가입
     *
     * @param signUpRequest 회원가입 요청 정보
     * @return String jwt 토큰
     */
    public TokenResponse signUp(SignUpRequest signUpRequest) {
        if (userRepository.existsByEmail(signUpRequest.email()))
            throw new CustomException(CustomErrorCode.ALREADY_EXIST_USER_ID);


        User savedUser = userRepository.save(User.Companion.create(
                signUpRequest.email(),
                passwordEncoder.encode(signUpRequest.password()),
                signUpRequest.name(),
                signUpRequest.phoneNumber()
        ));

        return new TokenResponse(
                jwtTokenProvider.createToken(savedUser.getId(), savedUser.getEmail(), savedUser.getRoles(), null)
        );
    }

    /**
     * 로그인
     *
     * @param loginRequest 로그인 요청 정보
     * @return User 엔티티
     */
    public TokenResponse login(LoginRequest loginRequest) {
        User user = userRepository.findByEmail(loginRequest.getEmail());

//        if (user.isEmpty()) throw new CustomException(CustomErrorCode.EMAIL_NOT_FOUND);

        if (!passwordEncoder.matches(loginRequest.getPassword(), user.getPassword()))
            throw new CustomException(CustomErrorCode.PASSWORD_NOT_MATCH);

        return new TokenResponse(
                jwtTokenProvider.createToken(user.getId(), user.getEmail(), user.getRoles(), null)
        );

    }

    public EmailResponse findEmailByPhone(String phone) {
        User user = userRepository.findByPhone(phone);
//        if (user.isEmpty()) throw new CustomException(CustomErrorCode.PHONE_NOT_FOUND);
        return new EmailResponse(user.getEmail());
    }

    /**
     * 아이디 중복 체크
     *
     * @param email 사용하려는 이메일
     * @return True: 사용 가능, False: 사용중인 이메일
     */
    public IdCheckResponse idCheck(String email) {
        return new IdCheckResponse(!userRepository.existsByEmail(email));
    }
}

File: src/main/kotlin/ac/kr/deu/connect/luck/auth/service/AuthService.kt
package ac.kr.deu.connect.luck.auth.service

import ac.kr.deu.connect.luck.auth.JwtTokenProvider
import ac.kr.deu.connect.luck.auth.controller.request.SignupRequest
import ac.kr.deu.connect.luck.common.exception.BadRequestException
import ac.kr.deu.connect.luck.common.exception.ConflictException
import ac.kr.deu.connect.luck.user.entity.User
import ac.kr.deu.connect.luck.user.repository.UserRepository
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class AuthService(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder,
    private val jwtTokenProvider: JwtTokenProvider,
) {

    /**
     * 회원가입
     * @param signupRequest 회원가입 요청 폼
     * @return Jwt 토큰
     */
    @Transactional
    fun signup(signupRequest: SignupRequest): String {
        if (userRepository.existsByEmail(signupRequest.email)) {
            throw ConflictException("Email already exists")
        }

        val user = userRepository.save(
            User.create(
                email = signupRequest.email,
                password = passwordEncoder.encode(signupRequest.password),
                name = signupRequest.name,
                phone = signupRequest.phoneNumber,
            )
        )

        return jwtTokenProvider.createToken(
            userId = user.id,
            email = user.email,
            roles = user.roles,
        )
    }

    /**
     * 로그인
     * @param email 이메일
     * @param password 비밀번호
     * @return Jwt 토큰
     */
    @Transactional(readOnly = true)
    fun login(email: String, password: String): String {
        val user = userRepository.findByEmail(email)
            ?: throw BadRequestException("이메일 혹은 비밀번호가 잘못되었습니다.")

        if (!passwordEncoder.matches(password, user.password)) {
            throw BadRequestException("이메일 혹은 비밀번호가 잘못되었습니다.")
        }

        return jwtTokenProvider.createToken(
            userId = user.id,
            email = user.email,
            roles = user.roles,
        )
    }

    /**
     * 이메일 사용가능 여부 확인
     * @param email 이메일
     * @return 사용가능 여부 (true: 사용가능, false: 중복)
     */
    @Transactional(readOnly = true)
    fun checkEmailDuplication(email: String): Boolean {
        return !userRepository.existsByEmail(email)
    }

    /**
     * 전화번호로 회원 조회
     * @param phone 전화번호
     * @return 사용자 이메일
     */
    @Transactional(readOnly = true)
    fun findEmailByPhone(phone: String): String {
        val user = userRepository.findByPhone(phone)
            ?: throw BadRequestException("전화번호로 사용자를 찾을 수 없습니다.")
        return user.email
    }
}
